NativeAOT on Alpine with StaticICULinking

I did try to use StaticICULinking on Alpine to make Docker from scratch more interesting
I add `clang libexecinfo binutils musl-dev build-base zlib-static cmake icu-static icu-dev` dependencies. Basically just add `cmake icu-static icu-dev` as additional packages.

Add following changes
```xml
<PropertyGroup>
   <InvariantGlobalization>false</InvariantGlobalization>
   <StaticICULinking>true</StaticICULinking>
</PropertyGroup>

<ItemGroup>
  <LinkerArg Include="-static" />
</ItemGroup>

<ItemGroup>
  <PackageReference Include="Microsoft.DotNet.ILCompiler" Version="8.0.0-alpha.1.22609.9" />
</ItemGroup>
```

And build fails with following LOOONG error message which I trim in preview, but attach as file
```
#13 13.35 /root/.nuget/packages/microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22609.9/build/Microsoft.DotNet.ILCompiler.SingleEntry.targets(45,5): warning : Delete explicit 'Microsoft.DotNet.ILCompiler' package reference in your project file. Explicit 'Microsoft.DotNet.ILCompiler' package reference can run into version errors. [/app/HelloWorldStatic.csproj]
#13 13.97   -- The C compiler identification is GNU 11.2.1
#13 13.99   -- Detecting C compiler ABI info
#13 14.11   -- Detecting C compiler ABI info - done
#13 14.14   -- Check for working C compiler: /usr/bin/cc - skipped
#13 14.14   -- Detecting C compile features
#13 14.14   -- Detecting C compile features - done
#13 14.15   -- Performing Test HAVE_UDAT_STANDALONE_SHORTER_WEEKDAYS
#13 14.38   -- Performing Test HAVE_UDAT_STANDALONE_SHORTER_WEEKDAYS - Success
#13 14.38   -- Looking for ucol_setMaxVariable
#13 14.68   -- Looking for ucol_setMaxVariable - found
#13 14.68   -- Looking for ucol_clone
#13 14.83   -- Looking for ucol_clone - found
#13 14.87   CMake Warning (dev) in CMakeLists.txt:
#13 14.87     No cmake_minimum_required command is present.  A line of code such as
#13 14.87
#13 14.87       cmake_minimum_required(VERSION 3.23)
#13 14.87
#13 14.87     should be added at the top of the file.  The version specified may be lower
#13 14.87     if you wish to support older CMake versions for this project.  For more
#13 14.87     information run "cmake --help-policy CMP0000".
#13 14.87   This warning is for project developers.  Use -Wno-dev to suppress it.
#13 14.87
#13 14.87   -- Configuring done
#13 14.87   -- Generating done
#13 14.87   -- Build files have been written to: /app/obj/Release/net7.0/linux-musl-x64/libs/System.Globalization.Native/build
#13 14.89   [  8%] Building C object CMakeFiles/System.Globalization.Native-Static.dir/pal_calendarData.o
#13 14.90   [ 25%] Building C object CMakeFiles/System.Globalization.Native-Static.dir/pal_casing.o
#13 14.90   [ 25%] Building C object CMakeFiles/System.Globalization.Native-Static.dir/pal_idna.o
#13 14.90   [ 33%] Building C object CMakeFiles/System.Globalization.Native-Static.dir/pal_localeNumberData.o
#13 14.90   [ 50%] Building C object CMakeFiles/System.Globalization.Native-Static.dir/pal_locale.o
#13 14.90   [ 50%] Building C object CMakeFiles/System.Globalization.Native-Static.dir/pal_localeStringData.o
#13 14.90   [ 58%] Building C object CMakeFiles/System.Globalization.Native-Static.dir/pal_collation.o
#13 14.91   [ 66%] Building C object CMakeFiles/System.Globalization.Native-Static.dir/pal_normalization.o
#13 14.91   [ 75%] Building C object CMakeFiles/System.Globalization.Native-Static.dir/pal_icushim_static.o
#13 14.92   [ 83%] Building C object CMakeFiles/System.Globalization.Native-Static.dir/pal_timeZoneInfo.o
#13 14.93   [ 91%] Building C object CMakeFiles/System.Globalization.Native-Static.dir/entrypoints.o
#13 15.26   /root/.nuget/packages/runtime.linux-musl-x64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22609.9/native/src/libs/System.Globalization.Native/pal_collation.c: In function 'CloneCollatorWithOptions':
#13 15.26 /root/.nuget/packages/runtime.linux-musl-x64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22609.9/native/src/libs/System.Globalization.Native/pal_collation.c(338,9): warning G46159011: 'ucol_safeClone_71' is deprecated [-Wdeprecated-declarations] [/app/HelloWorldStatic.csproj]
#13 15.26     338 |         pClonedCollator = ucol_safeClone(pCollator, NULL, NULL, pErr);
#13 15.26         |         ^~~~~~~~~~~~~~~
#13 15.26   In file included from /usr/include/unicode/platform.h:25,
#13 15.26                    from /usr/include/unicode/ptypes.h:52,
#13 15.26                    from /usr/include/unicode/umachine.h:46,
#13 15.26                    from /usr/include/unicode/utypes.h:38,
#13 15.26                    from /usr/include/unicode/ucurr.h:12,
#13 15.26                    from /root/.nuget/packages/runtime.linux-musl-x64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22609.9/native/src/libs/System.Globalization.Native/pal_icushim_internal.h:24,
#13 15.26                    from /root/.nuget/packages/runtime.linux-musl-x64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22609.9/native/src/libs/System.Globalization.Native/pal_errors_internal.h:6,
#13 15.26                    from /root/.nuget/packages/runtime.linux-musl-x64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22609.9/native/src/libs/System.Globalization.Native/pal_collation.c:11:
#13 15.26   /usr/include/unicode/ucol.h:1350:1: note: declared here
#13 15.26    1350 | ucol_safeClone(const UCollator *coll,
#13 15.26         | ^~~~~~~~~~~~~~
#13 15.55   /root/.nuget/packages/runtime.linux-musl-x64.microsoft.dotnet.ilcompiler/8.0.0-alpha.1.22609.9/native/src/libs/System.Globalization.Native/pal_collation.c: At top level:
#13 15.55   cc1: note: unrecognized command-line option '-Wno-unknown-warning-option' may have been intended to silence earlier diagnostics
#13 15.55   cc1: note: unrecognized command-line option '-Wno-extra-semi-stmt' may have been intended to silence earlier diagnostics
#13 15.55   cc1: note: unrecognized command-line option '-Wno-documentation-unknown-command' may have been intended to silence earlier diagnostics
#13 15.55   cc1: note: unrecognized command-line option '-Wno-documentation' may have been intended to silence earlier diagnostics
#13 15.55   cc1: note: unrecognized command-line option '-Wno-reserved-id-macro' may have been intended to silence earlier diagnostics
#13 15.55   cc1: note: unrecognized command-line option '-Wno-covered-switch-default' may have been intended to silence earlier diagnostics
#13 15.60   [100%] Linking C static library libSystem.Globalization.Native.a
#13 15.63   [100%] Built target System.Globalization.Native-Static
#13 15.67   Generating native code
#13 22.18 clang-13 : warning : argument unused during compilation: '-pie' [-Wunused-command-line-argument] [/app/HelloWorldStatic.csproj]
#13 22.35   /usr/bin/ld: attempted static link of dynamic object `/usr/bin/../lib/libc.so'
#13 22.72   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_longnames.ao): in function `icu_71::number::impl::MixedUnitLongNameHandler::forMeasureUnit(icu_71::Locale const&, icu_71::MeasureUnit const&, UNumberUnitWidth const&, char const*, icu_71::PluralRules const*, icu_71::number::impl::MicroPropsGenerator const*, icu_71::number::impl::MixedUnitLongNameHandler*, UErrorCode&)':
#13 22.72   number_longnames.cpp:(.text+0x2137): undefined reference to `__cxa_throw_bad_array_new_length'
#13 22.72   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_longnames.ao): in function `icu_71::number::impl::MixedUnitLongNameHandler::getMixedUnitModifier(icu_71::number::impl::DecimalQuantity&, icu_71::number::impl::MicroProps&, UErrorCode&) const':
#13 22.72   number_longnames.cpp:(.text+0x2918): undefined reference to `__cxa_throw_bad_array_new_length'
#13 22.72   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_longnames.ao): in function `icu_71::number::impl::LongNameMultiplexer::forMeasureUnits(icu_71::Locale const&, icu_71::MaybeStackVector<icu_71::MeasureUnit, 8> const&, UNumberUnitWidth const&, char const*, icu_71::PluralRules const*, icu_71::number::impl::MicroPropsGenerator const*, UErrorCode&)':
#13 22.72   number_longnames.cpp:(.text+0x4afe): undefined reference to `__cxa_throw_bad_array_new_length'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_longnames.ao):(.data.rel.ro._ZTIN6icu_716number4impl15LongNameHandlerE[_ZTIN6icu_716number4impl15LongNameHandlerE]+0x0): undefined reference to `vtable for __cxxabiv1::__vmi_class_type_info'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_longnames.ao):(.data.rel.ro._ZTIN6icu_716number4impl24MixedUnitLongNameHandlerE[_ZTIN6icu_716number4impl24MixedUnitLongNameHandlerE]+0x0): undefined reference to `vtable for __cxxabiv1::__vmi_class_type_info'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_longnames.ao):(.data.rel.ro._ZTIN6icu_716number4impl19LongNameMultiplexerE[_ZTIN6icu_716number4impl19LongNameMultiplexerE]+0x0): undefined reference to `vtable for __cxxabiv1::__vmi_class_type_info'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_longnames.ao):(.data.rel.ro+0x0): undefined reference to `vtable for __cxxabiv1::__si_class_type_info'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_longnames.ao):(.data.rel.ro+0x18): undefined reference to `vtable for __cxxabiv1::__si_class_type_info'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_modifiers.ao): in function `icu_71::number::impl::ConstantMultiFieldModifier::semanticallyEquivalent(icu_71::number::impl::Modifier const&) const':
#13 22.73   number_modifiers.cpp:(.text+0x2b8): undefined reference to `__dynamic_cast'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_modifiers.ao): in function `icu_71::number::impl::ConstantAffixModifier::semanticallyEquivalent(icu_71::number::impl::Modifier const&) const':
#13 22.73   number_modifiers.cpp:(.text+0x33e): undefined reference to `__dynamic_cast'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_modifiers.ao): in function `icu_71::number::impl::SimpleModifier::semanticallyEquivalent(icu_71::number::impl::Modifier const&) const':
#13 22.73   number_modifiers.cpp:(.text+0x3ac): undefined reference to `__dynamic_cast'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_modifiers.ao):(.data.rel.ro._ZTIN6icu_716number4impl8ModifierE[_ZTIN6icu_716number4impl8ModifierE]+0x0): undefined reference to `vtable for __cxxabiv1::__class_type_info'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_modifiers.ao):(.data.rel.ro._ZTIN6icu_716number4impl13ModifierStoreE[_ZTIN6icu_716number4impl13ModifierStoreE]+0x0): undefined reference to `vtable for __cxxabiv1::__class_type_info'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_modifiers.ao):(.data.rel.ro._ZTIN6icu_716number4impl21ConstantAffixModifierE[_ZTIN6icu_716number4impl21ConstantAffixModifierE]+0x0): undefined reference to `vtable for __cxxabiv1::__vmi_class_type_info'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_modifiers.ao):(.data.rel.ro._ZTIN6icu_716number4impl14SimpleModifierE[_ZTIN6icu_716number4impl14SimpleModifierE]+0x0): undefined reference to `vtable for __cxxabiv1::__vmi_class_type_info'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_modifiers.ao):(.data.rel.ro._ZTIN6icu_716number4impl26ConstantMultiFieldModifierE[_ZTIN6icu_716number4impl26ConstantMultiFieldModifierE]+0x0): undefined reference to `vtable for __cxxabiv1::__vmi_class_type_info'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_modifiers.ao):(.data.rel.ro._ZTIN6icu_716number4impl30CurrencySpacingEnabledModifierE[_ZTIN6icu_716number4impl30CurrencySpacingEnabledModifierE]+0x0): undefined reference to `vtable for __cxxabiv1::__si_class_type_info'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_modifiers.ao):(.data.rel.ro._ZTIN6icu_716number4impl21AdoptingModifierStoreE[_ZTIN6icu_716number4impl21AdoptingModifierStoreE]+0x0): undefined reference to `vtable for __cxxabiv1::__vmi_class_type_info'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_patternmodifier.ao):(.data.rel.ro._ZTIN6icu_716number4impl20AffixPatternProviderE[_ZTIN6icu_716number4impl20AffixPatternProviderE]+0x0): undefined reference to `vtable for __cxxabiv1::__class_type_info'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_patternmodifier.ao):(.data.rel.ro._ZTIN6icu_716number4impl24ImmutablePatternModifierE[_ZTIN6icu_716number4impl24ImmutablePatternModifierE]+0x0): undefined reference to `vtable for __cxxabiv1::__vmi_class_type_info'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_patternmodifier.ao):(.data.rel.ro._ZTIN6icu_716number4impl22MutablePatternModifierE[_ZTIN6icu_716number4impl22MutablePatternModifierE]+0x0): undefined reference to `vtable for __cxxabiv1::__vmi_class_type_info'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_scientific.ao): in function `icu_71::number::impl::ScientificModifier::semanticallyEquivalent(icu_71::number::impl::Modifier const&) const':
#13 22.73   number_scientific.cpp:(.text+0x5e): undefined reference to `__dynamic_cast'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_scientific.ao):(.data.rel.ro._ZTIN6icu_716number4impl18ScientificModifierE[_ZTIN6icu_716number4impl18ScientificModifierE]+0x0): undefined reference to `vtable for __cxxabiv1::__vmi_class_type_info'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(number_scientific.ao):(.data.rel.ro._ZTIN6icu_716number4impl17ScientificHandlerE[_ZTIN6icu_716number4impl17ScientificHandlerE]+0x0): undefined reference to `vtable for __cxxabiv1::__vmi_class_type_info'
#13 22.73   /usr/bin/ld: /usr/bin/../lib/libicuuc.a(ustr_titlecase_brkiter.ao):(.data.rel.ro._ZTIN6icu_7124WholeStringBreakIteratorE[_ZTIN6icu_7124WholeStringBreakIteratorE]+0x0): undefined reference to `vtable for __cxxabiv1::__si_class_type_info'
#13 22.79   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(collationbuilder.ao):(.data.rel.ro._ZTIN6icu_7116CollationBuilderE[_ZTIN6icu_7116CollationBuilderE]+0x0): undefined reference to `vtable for __cxxabiv1::__si_class_type_info'
#13 22.79   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(collationbuilder.ao):(.data.rel.ro+0x0): undefined reference to `vtable for __cxxabiv1::__si_class_type_info'
#13 22.79   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(collationbuilder.ao):(.data.rel.ro._ZTIN6icu_7111CEFinalizerE[_ZTIN6icu_7111CEFinalizerE]+0x0): undefined reference to `vtable for __cxxabiv1::__si_class_type_info'
#13 22.79   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(collationbuilder.ao):(.data.rel.local.DW.ref.__gxx_personality_v0[DW.ref.__gxx_personality_v0]+0x0): undefined reference to `__gxx_personality_v0'
#13 22.79   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(collationdatabuilder.ao):(.data.rel.ro._ZTIN6icu_7120CollationDataBuilder10CEModifierE[_ZTIN6icu_7120CollationDataBuilder10CEModifierE]+0x0): undefined reference to `vtable for __cxxabiv1::__si_class_type_info'
#13 22.79   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(collationdatabuilder.ao):(.data.rel.ro._ZTIN6icu_7120CollationDataBuilderE[_ZTIN6icu_7120CollationDataBuilderE]+0x0): undefined reference to `vtable for __cxxabiv1::__si_class_type_info'
#13 22.79   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(collationdatabuilder.ao):(.data.rel.ro._ZTIN6icu_7128DataBuilderCollationIteratorE[_ZTIN6icu_7128DataBuilderCollationIteratorE]+0x0): undefined reference to `vtable for __cxxabiv1::__si_class_type_info'
#13 22.80   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(collationfastlatinbuilder.ao):(.data.rel.ro._ZTIN6icu_7125CollationFastLatinBuilderE[_ZTIN6icu_7125CollationFastLatinBuilderE]+0x0): undefined reference to `vtable for __cxxabiv1::__si_class_type_info'
#13 22.80   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(collationiterator.ao):(.data.rel.ro._ZTIN6icu_7117CollationIteratorE[_ZTIN6icu_7117CollationIteratorE]+0x0): undefined reference to `vtable for __cxxabiv1::__si_class_type_info'
#13 22.80   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(collationruleparser.ao):(.data.rel.ro._ZTIN6icu_7119CollationRuleParser4SinkE[_ZTIN6icu_7119CollationRuleParser4SinkE]+0x0): more undefined references to `vtable for __cxxabiv1::__si_class_type_info' follow
#13 22.80   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(ucal.ao): in function `ucal_getDSTSavings_71':
#13 22.80   ucal.cpp:(.text+0x31f): undefined reference to `__dynamic_cast'
#13 22.80   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(ucal.ao): in function `ucal_setGregorianChange_71':
#13 22.80   ucal.cpp:(.text+0x51a): undefined reference to `__dynamic_cast'
#13 22.80   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(ucal.ao): in function `ucal_getGregorianChange_71':
#13 22.80   ucal.cpp:(.text+0x59a): undefined reference to `__dynamic_cast'
#13 22.80   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(ucal.ao): in function `ucal_getTimeZoneOffsetFromLocal_71':
#13 22.80   ucal.cpp:(.text+0xdf5): undefined reference to `__dynamic_cast'
#13 22.80   /usr/bin/ld: /usr/bin/../lib/libicui18n.a(ucal.ao): in function `ucal_getTimeZoneTransitionDate_71':
#13 22.80   ucal.cpp:
(.text+0x1252): undefined reference to `__dynamic_cast'
```

Any ideas what am I missing?


[report.txt](https://github.com/dotnet/runtime/files/10200665/report.txt)

--- a/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets
+++ b/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets
@@ -46,10 +46,10 @@ The .NET Foundation licenses this file to you under the MIT license.
 
     <ItemGroup>
       <NativeLibrary Condition="'$(IlcMultiModule)' == 'true'" Include="$(SharedLibrary)" />
-      <NativeLibrary Condition="$(NativeLib) == ''" Include="$(IlcSdkPath)libbootstrapper.a" />
-      <NativeLibrary Condition="$(NativeLib) != ''" Include="$(IlcSdkPath)libbootstrapperdll.a" />
+      <NativeLibrary Condition="'$(NativeLib)' == ''" Include="$(IlcSdkPath)libbootstrapper.a" />
+      <NativeLibrary Condition="'$(NativeLib)' != ''" Include="$(IlcSdkPath)libbootstrapperdll.a" />
       <NativeLibrary Include="$(IlcSdkPath)$(FullRuntimeName).a" />
-      <NativeLibrary Condition="'$(LinkStandardCPlusPlusLibrary)' != 'true'" Include="$(IlcSdkPath)libstdc++compat.a" />
+      <NativeLibrary Condition="'$(LinkStandardCPlusPlusLibrary)' != 'true' and '$(StaticICULinking)' != 'true'" Include="$(IlcSdkPath)libstdc++compat.a" />
     </ItemGroup>
 
     <ItemGroup>
@@ -72,7 +72,8 @@ The .NET Foundation licenses this file to you under the MIT license.
     <ItemGroup Condition="'$(StaticICULinking)' == 'true'">
       <NativeLibrary Include="$(IntermediateOutputPath)/libs/System.Globalization.Native/build/libSystem.Globalization.Native.a"/>
       <DirectPInvoke Include="libSystem.Globalization.Native" />
-      <StaticICULibs Include="-Wl,-Bstatic -licuio -licutu -licui18n -licuuc -licudata -Wl,-Bdynamic" />
+      <StaticICULibs Include="-Wl,-Bstatic -licuio -licutu -licui18n -licuuc -licudata -lstdc++" />
+      <StaticICULibs Include="-Wl,-Bdynamic" Condition="'$(NativeLib)' != 'Static' and !@(LinkerArg->Contains('-static'))" />
     </ItemGroup>
 
     <ItemGroup Condition="'$(TargetOS)' == 'OSX'">

--- a/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets
+++ b/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets
@@ -72,8 +72,9 @@ The .NET Foundation licenses this file to you under the MIT license.
     <ItemGroup Condition="'$(StaticICULinking)' == 'true'">
       <NativeLibrary Include="$(IntermediateOutputPath)/libs/System.Globalization.Native/build/libSystem.Globalization.Native.a"/>
       <DirectPInvoke Include="libSystem.Globalization.Native" />
-      <StaticICULibs Include="-Wl,-Bstatic -licuio -licutu -licui18n -licuuc -licudata -lstdc++" />
-      <StaticICULibs Include="-Wl,-Bdynamic" Condition="'$(NativeLib)' != 'Static' and !@(LinkerArg->Contains('-static'))" />
+      <StaticICULibs Include="-Wl,-Bstatic" Condition="'$(NativeLib)' != 'Static' and '@(LinkerArg->Contains('-static'))' != 'true'" />
+      <StaticICULibs Include="-licuio -licutu -licui18n -licuuc -licudata -lstdc++" />
+      <StaticICULibs Include="-Wl,-Bdynamic" Condition="'$(NativeLib)' != 'Static' and '@(LinkerArg->Contains('-static'))' != 'true'" />
     </ItemGroup>
 
     <ItemGroup Condition="'$(TargetOS)' == 'OSX'">

--- a/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets
+++ b/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets
@@ -20,7 +20,6 @@ The .NET Foundation licenses this file to you under the MIT license.
     <CppLinker>$(CppCompilerAndLinker)</CppLinker>
     <CppLibCreator>ar</CppLibCreator>
     <DsymUtilOptions Condition="'$(TargetOS)' == 'OSX'">--flat</DsymUtilOptions>
-    <StaticExecutable Condition="'$(StaticExecutable)' == '' and '@(LinkerArg->Contains('-static'))' == 'true'">true</StaticExecutable>
   </PropertyGroup>
 
   <Target Name="SetupOSSpecificProps" DependsOnTargets="$(IlcDynamicBuildPropertyDependencies)">

--- a/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets
+++ b/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets
@@ -20,6 +20,7 @@ The .NET Foundation licenses this file to you under the MIT license.
     <CppLinker>$(CppCompilerAndLinker)</CppLinker>
     <CppLibCreator>ar</CppLibCreator>
     <DsymUtilOptions Condition="'$(TargetOS)' == 'OSX'">--flat</DsymUtilOptions>
+    <StaticExecutable Condition="'$(StaticExecutable)' == '' and '@(LinkerArg->Contains('-static'))' == 'true'">true</StaticExecutable>
   </PropertyGroup>
 
   <Target Name="SetupOSSpecificProps" DependsOnTargets="$(IlcDynamicBuildPropertyDependencies)">
@@ -69,12 +70,12 @@ The .NET Foundation licenses this file to you under the MIT license.
       <NativeLibrary Include="@(NetCoreAppNativeLibrary->'%(EscapedPath)')" />
     </ItemGroup>
 
-    <ItemGroup Condition="'$(StaticICULinking)' == 'true'">
+    <ItemGroup Condition="'$(StaticICULinking)' == 'true' and '$(NativeLib)' != 'Static'">
       <NativeLibrary Include="$(IntermediateOutputPath)/libs/System.Globalization.Native/build/libSystem.Globalization.Native.a"/>
       <DirectPInvoke Include="libSystem.Globalization.Native" />
-      <StaticICULibs Include="-Wl,-Bstatic" Condition="'$(NativeLib)' != 'Static' and '@(LinkerArg->Contains('-static'))' != 'true'" />
+      <StaticICULibs Include="-Wl,-Bstatic" Condition="'$(StaticExecutable)' != 'true'" />
       <StaticICULibs Include="-licuio -licutu -licui18n -licuuc -licudata -lstdc++" />
-      <StaticICULibs Include="-Wl,-Bdynamic" Condition="'$(NativeLib)' != 'Static' and '@(LinkerArg->Contains('-static'))' != 'true'" />
+      <StaticICULibs Include="-Wl,-Bdynamic" Condition="'$(StaticExecutable)' != 'true'" />
     </ItemGroup>
 
     <ItemGroup Condition="'$(TargetOS)' == 'OSX'">
@@ -88,6 +89,7 @@ The .NET Foundation licenses this file to you under the MIT license.
     <Exec Command="$(IlcHostPackagePath)/native/src/libs/System.Globalization.Native/local_build.sh $(IlcHostPackagePath)/ $(IntermediateOutputPath)" Condition="'$(StaticICULinking)' == 'true'"/>
 
     <ItemGroup>
+      <LinkerArg Include="-static" Condition="'$(StaticExecutable)' == 'true'" />
       <LinkerArg Include="@(NativeLibrary)" />
       <LinkerArg Include="--sysroot=$(SysRoot)" Condition="'$(SysRoot)' != ''" />
       <LinkerArg Include="--target=$(TargetTriple)" Condition="'$(TargetOS)' != 'OSX' and '$(TargetTriple)' != ''" />

--- a/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Windows.targets
+++ b/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Windows.targets
@@ -21,7 +21,7 @@ The .NET Foundation licenses this file to you under the MIT license.
     <FullRuntimeName>Runtime.WorkstationGC</FullRuntimeName>
     <FullRuntimeName Condition="'$(ServerGarbageCollection)' == 'true' or '$(ControlFlowGuard)' == 'Guard'">Runtime.ServerGC</FullRuntimeName>
     <BootstrapperName>bootstrapper</BootstrapperName>
-    <BootstrapperName Condition="$(NativeLib) != ''">bootstrapperdll</BootstrapperName>
+    <BootstrapperName Condition="'$(NativeLib)' != ''">bootstrapperdll</BootstrapperName>
     <EntryPointSymbol Condition="'$(EntryPointSymbol)' == ''">wmainCRTStartup</EntryPointSymbol>
     <LinkerSubsystem Condition="'$(OutputType)' == 'WinExe' and '$(LinkerSubsystem)' == ''">WINDOWS</LinkerSubsystem>
     <LinkerSubsystem Condition="'$(OutputType)' == 'Exe' and '$(LinkerSubsystem)' == ''">CONSOLE</LinkerSubsystem>
@@ -72,7 +72,7 @@ The .NET Foundation licenses this file to you under the MIT license.
     </ItemGroup>
 
     <ItemGroup>
-      <LinkerArg Condition="$(NativeLib) == 'Shared'" Include="/DLL" />
+      <LinkerArg Condition="'$(NativeLib)' == 'Shared'" Include="/DLL" />
       <LinkerArg Include="@(NativeLibrary->'&quot;%(Identity)&quot;')" />
       <LinkerArg Include="@(SdkNativeLibrary->'&quot;%(Identity)&quot;')" />
       <LinkerArg Include="/NOLOGO /MANIFEST:NO" />
@@ -81,7 +81,7 @@ The .NET Foundation licenses this file to you under the MIT license.
       <LinkerArg Include="/INCREMENTAL:NO" />
       <LinkerArg Condition="'$(LinkerSubsystem)' != ''" Include="/SUBSYSTEM:$(LinkerSubsystem)" />
       <LinkerArg Condition="'$(OutputType)' == 'WinExe' or '$(OutputType)' == 'Exe'" Include="/ENTRY:$(EntryPointSymbol) /NOEXP /NOIMPLIB" />
-      <LinkerArg Condition="$(NativeLib) == 'Shared'" Include="/INCLUDE:NativeAOT_StaticInitialization" />
+      <LinkerArg Condition="'$(NativeLib)' == 'Shared'" Include="/INCLUDE:NativeAOT_StaticInitialization" />
       <LinkerArg Include="/NATVIS:&quot;$(MSBuildThisFileDirectory)NativeAOT.natvis&quot;" />
       <LinkerArg Condition="'$(ControlFlowGuard)' == 'Guard'" Include="/guard:cf" />
     </ItemGroup>

--- a/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets
+++ b/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets
@@ -67,11 +67,11 @@ The .NET Foundation licenses this file to you under the MIT license.
 
     <NativeBinaryExt Condition="'$(IsNativeExecutable)' == 'true' and '$(TargetOS)' == 'windows'">.exe</NativeBinaryExt>
     <NativeBinaryExt Condition="'$(IsNativeExecutable)' == 'true' and '$(TargetOS)' != 'windows'"></NativeBinaryExt>
-    <NativeBinaryExt Condition="'$(IsNativeExecutable)' != 'true' and '$(TargetOS)' == 'windows' and $(NativeLib) == 'Shared'">.dll</NativeBinaryExt>
-    <NativeBinaryExt Condition="'$(IsNativeExecutable)' != 'true' and '$(TargetOS)' == 'OSX' and $(NativeLib) == 'Shared'">.dylib</NativeBinaryExt>
-    <NativeBinaryExt Condition="'$(IsNativeExecutable)' != 'true' and '$(TargetOS)' != 'windows' and '$(TargetOS)' != 'OSX' and $(NativeLib) == 'Shared'">.so</NativeBinaryExt>
-    <NativeBinaryExt Condition="'$(IsNativeExecutable)' != 'true' and '$(TargetOS)' == 'windows' and $(NativeLib) == 'Static'">.lib</NativeBinaryExt>
-    <NativeBinaryExt Condition="'$(IsNativeExecutable)' != 'true' and '$(TargetOS)' != 'windows' and $(NativeLib) == 'Static'">.a</NativeBinaryExt>
+    <NativeBinaryExt Condition="'$(IsNativeExecutable)' != 'true' and '$(TargetOS)' == 'windows' and '$(NativeLib)' == 'Shared'">.dll</NativeBinaryExt>
+    <NativeBinaryExt Condition="'$(IsNativeExecutable)' != 'true' and '$(TargetOS)' == 'OSX' and '$(NativeLib)' == 'Shared'">.dylib</NativeBinaryExt>
+    <NativeBinaryExt Condition="'$(IsNativeExecutable)' != 'true' and '$(TargetOS)' != 'windows' and '$(TargetOS)' != 'OSX' and '$(NativeLib)' == 'Shared'">.so</NativeBinaryExt>
+    <NativeBinaryExt Condition="'$(IsNativeExecutable)' != 'true' and '$(TargetOS)' == 'windows' and '$(NativeLib)' == 'Static'">.lib</NativeBinaryExt>
+    <NativeBinaryExt Condition="'$(IsNativeExecutable)' != 'true' and '$(TargetOS)' != 'windows' and '$(NativeLib)' == 'Static'">.a</NativeBinaryExt>
 
     <NativeSymbolExt Condition="'$(TargetOS)' == 'OSX'">.dwarf</NativeSymbolExt>
     <NativeSymbolExt Condition="'$(TargetOS)' == 'windows'">.pdb</NativeSymbolExt>
@@ -82,7 +82,7 @@ The .NET Foundation licenses this file to you under the MIT license.
 
     <NativeObject>$(NativeIntermediateOutputPath)$(TargetName)$(NativeObjectExt)</NativeObject>
     <NativeBinary>$(NativeOutputPath)$(TargetName)$(NativeBinaryExt)</NativeBinary>
-    <IlcExportUnmanagedEntrypoints Condition="$(NativeLib) == 'Shared'">true</IlcExportUnmanagedEntrypoints>
+    <IlcExportUnmanagedEntrypoints Condition="'$(NativeLib)' == 'Shared'">true</IlcExportUnmanagedEntrypoints>
     <ExportsFile Condition="$(IlcExportUnmanagedEntrypoints) == 'true' and $(ExportsFile) == ''">$(NativeIntermediateOutputPath)$(TargetName)$(ExportsFileExt)</ExportsFile>
 
     <IlcCompileOutput>$(NativeObject)</IlcCompileOutput>
@@ -225,7 +225,7 @@ The .NET Foundation licenses this file to you under the MIT license.
       <IlcArg Condition="$(IlcGenerateDgmlFile) == 'true'" Include="--scandgmllog:$(NativeIntermediateOutputPath)%(ManagedBinary.Filename).scan.dgml.xml" />
       <IlcArg Include="@(RdXmlFile->'--rdxml:%(FullPath)')" />
       <IlcArg Include="@(TrimmerRootDescriptor->'--descriptor:%(FullPath)')" />
-      <IlcArg Condition="$(NativeLib) != ''" Include="--nativelib" />
+      <IlcArg Condition="'$(NativeLib)' != ''" Include="--nativelib" />
       <IlcArg Condition="$(ExportsFile) != ''" Include="--exportsfile:$(ExportsFile)" />
       <IlcArg Include="@(AutoInitializedAssemblies->'--initassembly:%(Identity)')" />
       <IlcArg Include="@(RuntimeHostConfigurationOption->'--appcontextswitch:%(Identity)=%(Value)')" />

